ibtk/src/math/HierarchyMathOps.cpp

```diff
@@ -2033,14 +2064,32 @@ HierarchyMathOps::interp(const int dst_idx,
 {
     if (src_ghost_fill) src_ghost_fill->fillData(src_ghost_fill_time);
 
+    if (dst_cf_bdry_synch && dst_var->fineBoundaryRepresentsVariable() == false)
+    {
+        TBOX_ERROR("The combination of a destination boundary sync and coarse values representing the values on "
+                   "coarse-fine interfaces is not implemented.");
+    }
```

Comment:
There is no operator for outernode data corresponding to this setup so lets just abort.

Reply:
Are you sure that this doesn't work? I think that this is what is done in the side-to-node curl implementation.

My preference would be always to coarsen if `dst_cf_bdry_synch == true`.

Reply:
I just double-checked: the side-to-node curl implementation seems to work fine with either value of `fine_boundary_represents_variable`. I looked at the visual output and also ran the test suite.

Reply:
It does but that's because SAMRAI (and by extension us) always ignore `fine_boundary_represents_variable`. My reasoning here is that if that is `true` then synchronization should be coarsening (set coarse values with fine), whereas if it is `false` then we should use refinement (set fine with coarse). In both `curl` and `interp` we use restriction (`xeqScheduleOuternodeRestriction`), which works since SAMRAI provides the class `OuternodeDoubleConstantCoarsen`. However: there is no `OuternodeDoubleConstantRefine` so the other way isn't implemented.

Reply:
We can change the behavior for the sake of convenience but I think this is the correct setup in terms of "this means coarsening" and which outernode operators are implemented right now.

Reply:
I do not agree. We should only use the coarse grid interpolation to set the coarse grid values along the coarse-fine interface if the user says not to synchronize (`dst_cf_bdry_synch == false`). The reason is that the coarse grid interpolation will be using "invalid" values, whereas the fine grid interpolation will be using "valid" values.

If we want to synchronize coarse-fine interface values, we should always determine the overlying coarse grid values from the underlying fine grid values.

If a caller wants to do something different, they are able to do so by setting `dst_cf_bdry_synch == false` and then doing what they want along the coarse-fine interface.

Similarly, if a caller indicates that they want to use the coarse grid values in the `NodeVariable` to determine the value of the variable, then after setting the coarse grid values, we should go back and force the underling fine grid values in the `NodeVariable` to correspond to the coarse grid values. As above, if the caller wants to do something different (although I am having trouble seeing a use case), they are able to do so by setting `dst_cf_bdry_synch == false` and then doing what they want along the coarse-fine interface.

Reply:
I'm missing something important here - the coarse-grid side-to-node interpolation is not done with the outernode synchronization step, correct? Where does that part happen/where should that part happen?

Reply:
One way to think about this is that the side-to-node interpolation first happens patch-by-patch on each level of the patch hierarchy, independent of values on other patch levels. We then have to decide what (if anything) to do at coarse-fine interfaces and in the "invalid" regions of coarser levels.

My thinking in these operators is that if we view a particular variable as a "composite grid variable", we want to use the "valid" values to determine coarse-grid values. So generally, at overlaps between levels, we want to define coarse values in terms of overlying fine values.

If `dst_cf_bdry_synch == true`, then I think that the coarse values along the coarse-fine interface should be determined from the overlying fine values. That "synchronizes" the coarse values along the coarse-fine interface.

In addition, if `dst_cf_bdry_synch == true` and `fineBoundaryRepresentsVariable() == false`, then that says that ***in the composite grid variable***, the coarse values should determine the values of the overlying fine values along the coarse-fine interface. Ensuring that requires applying the "H1 fixup" along the coarse-fine interface.

Does this make sense?

Reply:
One other thing: here, I don't think that `dst_var->fineBoundaryRepresentsVariable() == false` means that the coarse-grid node data should be determined from the coarse-grid side data. Instead, I think that `dst_var->fineBoundaryRepresentsVariable() == false` simply implies that along the coarse-fine interface, the fine values should be determined from the overlying coarse values, e.g., through the H1 fixup routine.

Reply:
> You're probably right but I want to make sure I understand what's going on.
> 
> > If we want to synchronize coarse-fine interface values, we should always determine the overlying coarse grid values from the underlying fine grid values.
> 
> I'm stuck at this step - isn't this the outernode restriction which we presently use? If so then doesn't this step always imply that `fine_boundary_represents_variable` must be true?

As far as I can tell, this flag is only used in SAMRAI for refining and not for coarsening.

Reply:
To summarize:

If `dst_cf_bdry_synch == false`, then we simply apply the patch-wise interpolation routine on every patch in the hierarchy, with no interlevel communication.

If `dst_cf_bdry_synch == true`, then we always restrict the results from finer levels onto coarser levels along the coarse-fine interface (via `xeqScheduleOuternodeRestriction`), independent of the value of `dst_var->fineBoundaryRepresentsVariable()`.

If `dst_cf_bdry_synch == true && dst_var->fineBoundaryRepresentsVariable() == false`, then we also apply the H1 fixup routine, to ensure that the fine coarse-fine interface values interpolate the overlying coarse-grid values. We consider this part of the "synchronization process" along coarse-fine interfaces.

Reply:
Specifically, if `dst_var` is configured so that `fineBoundaryRepresentsVariable() == true`, we interpret that as an indication that the caller ***does not*** want to constrain the fine-grid values along the coarse-fine interface to interpolate the overlying coarse-grid values.

Reply:
Finally, notice that the behavior here can be customized by setting `dst_cf_bdry_synch = false` and then applying a custom synchronization routine after the call to `HierarchyMathOps::interp()`.

Reply:
(One reason why it is actually ***good*** that SAMRAI ignores the result of `fineBoundaryRepresentsVariable()` in coarsening is that the only way to perform composite-grid calculations in SAMRAI that use both fine and coarse DOFs is on the ***fine patches*** along the coarse-fine interface; the result can then be communicated to the coarse patch through a coarsening operator.)

Reply:
We chatted about this over zoom and the conclusion is:

1. if "synchronize dst" = false then we don't do anything with outernode or the H1 fixup routine
2. if "synchronize dst" = true and "fine side represents the solution" = false then we do outernode coarsening and the H1 fixup routine
3. if "synchronize dst" = true and "fine side represents the solution" = false then we do outernode coarsening (but no H1 fixup)